common/parse.go

```diff
@@ -114,66 +118,171 @@ func (jsonManager) ArrToMap(arr []*ajson.Node) ([]map[string]any, error) {
+		return nil, ErrNotNumeric
 	}
 
 	if math.Mod(count, 1.0) != 0 {
-		return 0, ErrNotInteger
+		return nil, ErrNotInteger
 	}
 
-	return int64(count), nil
+	result := int64(count)
+
+	return &result, nil
 }
 
 func (jsonManager) GetArr(node *ajson.Node, key string) ([]*ajson.Node, error) {
+	if !node.HasKey(key) {
+		return nil, createKeyNotFoundErr(key)
+	}
+
 	records, err := node.GetKey(key)
 	if err != nil {
 		return nil, err
 	}
 
 	arr, err := records.GetArray()
 	if err != nil {
-		return nil, ErrNotArray
+		return nil, formatProblematicKeyError(key, ErrNotArray)
 	}
 
 	return arr, nil
 }
 
-func (jsonManager) ArrSize(node *ajson.Node, key string) (int64, error) {
-	innerNode, err := node.GetKey(key)
+func (m jsonManager) ArrSize(node *ajson.Node, keys string) (int64, error) {
+	arr, err := m.GetArr(node, keys)
 	if err != nil {
 		return 0, err
 	}
 
-	if !innerNode.IsArray() {
-		return 0, ErrNotArray
+	return int64(len(arr)), nil
+}
+
+func (m jsonManager) GetStringWithDefault(node *ajson.Node, key string, defaultValue string) (string, error) {
+	result, err := m.GetString(node, key, true)
+	if err != nil {
+		return "", err
+	}
+
+	if result == nil {
+		return defaultValue, err
 	}
 
-	return int64(innerNode.Size()), nil
+	return *result, nil
 }
 
-func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (string, error) {
-	if optional && !node.HasKey(key) {
-		// null value in payload is allowed
-		return "", nil
+func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (*string, error) {
+	if !node.HasKey(key) {
+		if optional {
+			// null value in payload is allowed
+			return nil, nil // nolint:nilnil
+		} else {
+			return nil, createKeyNotFoundErr(key)
+		}
 	}
 
 	innerNode, err := node.GetKey(key)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+
+	if innerNode.IsNull() {
+		return nil, handleNullNode(key, optional)
 	}
 
 	txt, err := innerNode.GetString()
 	if err != nil {
-		return "", ErrNotString
+		return nil, ErrNotString
+	}
+
+	return &txt, nil
+}
+
+// GetNestedObject reaches into the JSON node by zooming keys
+// Ex: keys = [item, shipping, address] => item has object with shipping key and so on.
+func (jsonManager) GetNestedObject(node *ajson.Node, keys ...string) (*ajson.Node, error) {
+	var err error
+
+	// traverse nested JSON, use every key to zoom in
+	for _, key := range keys {
+		if !node.HasKey(key) {
+			message := fmt.Sprintf("%v; zoom=%v", key, strings.Join(keys, " "))
+
+			return nil, createKeyNotFoundErr(message)
+		}
+
+		node, err = node.GetKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !node.IsObject() {
+		return nil, ErrNotObject
+	}
+
+	return node, nil
+}
+
+func (jsonManager) ObjToMap(node *ajson.Node) (map[string]any, error) {
+	data, err := node.GetObject()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]any)
+	for k, v := range data {
+		result[k] = v
 	}
 
-	return txt, nil
+	return result, nil
```

Comment:
Use `node.Unpack` instead -- that will actually turn them in to go native types. Assuming that's your goal.

Reply:
This is great. It makes response better for the end user of this library.
I previously had some issues with tests because of it. With this change tests will be better and common.WriteResult will have golang types which is a heavy plus.